src/charm.py

```diff
@@ -291,6 +308,131 @@ def _on_db_router_relation_changed(self, event: RelationChangedEvent) -> None:
 
         self.unit.status = ActiveStatus()
 
+    def _on_shared_db_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy shared_db relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        if not self.unit.is_leader():
+            return
+
+        self.unit.status = MaintenanceStatus("Setting up shared-db relation")
+        logger.warning("DEPRECATION WARNING - `shared-db` is a legacy interface")
+
+        unit_relation_databag = event.relation.data[self.unit]
+        app_relation_databag = event.relation.data[self.app]
+
+        if unit_relation_databag.get("password"):
+            # Test if relation data is already set
+            # and avoid overwriting it
+            logger.warning("Data for shared-db already set.")
+            self.unit.status = ActiveStatus()
+            return
+
+        # retrieve data from the relation databag
+        # Abstracted is the fact that it is received from the leader unit of the related app
+        requires_relation_databag = event.relation.data[event.unit]
+        database_name = requires_relation_databag.get("database")
+        database_user = requires_relation_databag.get("username")
+
+        if not database_name or not database_user:
+            # Cannot create scoped database without credentials
+            logger.warning(
+                "Missing information for shared-db relation to create a database and scoped user"
+            )
+            self.unit.status = WaitingStatus("Missing information for shared-db relation")
+            return
+
+        password = generate_random_password(PASSWORD_LENGTH)
+
+        try:
+            self._mysql.create_application_database_and_scoped_user(
+                database_name, database_user, password
+            )
+
+            # set the relation data for consumption
+            cluster_primary = self._mysql.get_cluster_primary_address()
+
+            unit_relation_databag["db_host"] = cluster_primary.split(":")[0]
+            # Database port is static in legacy charm
+            unit_relation_databag["db_port"] = "3306"
+            # Wait timeout is a config option in legacy charm
+            # defaulted to 3600 seconds
+            unit_relation_databag["wait_timeout"] = "3600"
+            unit_relation_databag["password"] = password
```

Comment:
soemthing to confirm using jhack - are the values in the databag json encoded? they were for `db-router` (strings were `"string_value"` instead of `string_value`)

Reply:
I've noticed that while investigating, and `shared-db` does not json encodes it - so I choose to keep it *fac-simile* from the original one.